src/lib/storage/abstract_pos_list.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "types.hpp"
+#include "utils/assert.hpp"
+#include "utils/performance_warning.hpp"
+
+namespace opossum {
+
+class PosList;
+
+class AbstractPosList {
+ public:
+  template <typename PosListType = const AbstractPosList*, typename DereferenceReturnType = RowID>
+  class PosListIterator : public boost::iterator_facade<PosListIterator<PosListType, DereferenceReturnType>, RowID,
+                                                        boost::random_access_traversal_tag, DereferenceReturnType> {
+   public:
+    PosListIterator(PosListType pl_pointer, ChunkOffset pos, ChunkOffset max_size) {
+      _pl_pointer = pl_pointer;
+      _chunk_offset = pos;
+      _max_size = max_size;
+    }
+
+    // boost will not use the random_access_iterator_tag if reference_type is not a c++ reference (which it isn't here)
+    // we still want to use random access (for binary search, distance, ...)
+    typedef std::random_access_iterator_tag iterator_category;
+
+    PosListIterator() = delete;
+
+    void increment() { ++_chunk_offset; }
+
+    void decrement() { --_chunk_offset; }
+
+    void advance(std::ptrdiff_t n) { _chunk_offset += n; }
+
+    bool equal(const PosListIterator& other) const {
+      DebugAssert(_pl_pointer == other._pl_pointer, "You trusted me and I failed you.");
+      return other._chunk_offset == _chunk_offset;
+    }
+
+    std::ptrdiff_t distance_to(const PosListIterator& other) const {
+      return static_cast<std::ptrdiff_t>(other._chunk_offset) - _chunk_offset;
+    }
+
+    DereferenceReturnType dereference() const {
+      DebugAssert(_chunk_offset < _max_size, "You foool");
+      return (*_pl_pointer)[_chunk_offset];
+    }
+
+    PosListType _pl_pointer;
+    ChunkOffset _chunk_offset;
+    ChunkOffset _max_size;
+  };
+
+  // public:
+  virtual ~AbstractPosList() = default;
+
+  AbstractPosList& operator=(AbstractPosList&& other) = default;
+
+  // Returns whether it is guaranteed that the PosList references a single ChunkID.
+  // However, it may be false even if this is the case.
+  virtual bool references_single_chunk() const = 0;
+
+  // For chunks that share a common ChunkID, returns that ID.
+  virtual ChunkID common_chunk_id() const = 0;
+
+  virtual RowID operator[](size_t n) const = 0;
+
+  PosListIterator<> begin() const {
```

Comment:
Done for the 4 iterator methods (`begin`, `end`, `cbegin`, `cend`). The `PosListIterator` methods have to stay here because of templating and the `inline operator==` implementation has to stay here so it's available in every compilation unit, right?

Should we also do that for the implementations of the methods in matches_all_pos_list.hpp / pos_list.hpp (old name)?

Should we benchmark the performance before / after moving? I wouldn't expect these methods to be performance critical.

Reply:
As we don't use link-time optimization (because it was not beneficial so far and takes a lot of time), I'd expect that having things like `operator[]` in a separate class is bad for performance. Feel free to have look, but for me it would be sufficient to move (c)begin/c(end)/memory_usage in the other classes.

Reply:
Moved out all the methods that were not just forwarding the call and not operator[].